pkg/controller/daemon/manager.go

```diff
@@ -0,0 +1,475 @@
+	dcKey, err := controller.KeyFunc(dc)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+		return
+	}
+	dm.expectations.SetExpectations(dcKey, len(nodesNeedingDaemons), len(podsToDelete))
+
+	glog.V(4).Infof("Nodes needing daemons for daemon %s: %+v", dc.Name, nodesNeedingDaemons)
+	for i := range nodesNeedingDaemons {
+		if err := dm.podControl.CreateReplicaOnNode(dc.Namespace, dc, nodesNeedingDaemons[i]); err != nil {
+			glog.V(2).Infof("Failed creation, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.CreationObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+
+	glog.V(4).Infof("Pods to delete for daemon %s: %+v", dc.Name, podsToDelete)
+	for i := range podsToDelete {
+		if err := dm.podControl.DeletePod(dc.Namespace, podsToDelete[i]); err != nil {
+			glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q/%q", dc.Namespace, dc.Name)
+			dm.expectations.DeletionObserved(dcKey)
+			util.HandleError(err)
+		}
+	}
+}
+
+func storeDaemonStatus(dcClient client.DaemonInterface, dc *expapi.Daemon, desiredNumberScheduled, currentNumberScheduled, numberMisscheduled int) error {
+	if dc.Status.DesiredNumberScheduled == desiredNumberScheduled && dc.Status.CurrentNumberScheduled == currentNumberScheduled && dc.Status.NumberMisscheduled == numberMisscheduled {
+		return nil
+	}
+
+	var updateErr, getErr error
+	for i := 0; i <= StatusUpdateRetries; i++ {
+		dc.Status.DesiredNumberScheduled = desiredNumberScheduled
+		dc.Status.CurrentNumberScheduled = currentNumberScheduled
+		dc.Status.NumberMisscheduled = numberMisscheduled
+		_, updateErr := dcClient.Update(dc)
+		if updateErr == nil {
+			return updateErr
+		}
+		// Update the controller with the latest resource version for the next poll
+		if dc, getErr = dcClient.Get(dc.Name); getErr != nil {
+			// If the GET fails we can't trust status.Replicas anymore. This error
+			// is bound to be more interesting than the update failure.
+			return getErr
+		}
+	}
+	// Failed 2 updates one of which was with the latest controller, return the update error
+	return updateErr
+}
+
+func (dm *DaemonManager) updateDaemonStatus(dc *expapi.Daemon) {
+	glog.Infof("Updating daemon status")
+	nodeToDaemonPods, err := dm.getNodesToDaemonPods(dc)
+	if err != nil {
+		glog.Errorf("Error getting node to daemon pod mapping for daemon %+v: %v", dc, err)
+	}
+
+	nodeList, err := dm.nodeStore.List()
+	if err != nil {
+		glog.Errorf("Couldn't get list of nodes when adding daemon %+v: %v", dc, err)
+	}
+
+	var desiredNumberScheduled, currentNumberScheduled, numberMisscheduled int
+	for i := range nodeList.Items {
+		nodeSelector := labels.Set(dc.Spec.Template.Spec.NodeSelector).AsSelector()
+		shouldRun := nodeSelector.Matches(labels.Set(nodeList.Items[i].Labels))
+		numDaemonPods := len(nodeToDaemonPods[nodeList.Items[i].Name])
+		if shouldRun {
```

Comment:
I think that's better according to the API comments but we should increment by only one when numDaemonPods > 0:
```golang
if numDaemonPods > 0 {
        currentNumberScheduled++
}

if shouldRun {
        desiredNumberScheduled++
} else if numDaemonPods >= 1 {
        numberMisscheduled++
}
```